### PR TITLE
[L0] fix implementation of device-related program queries

### PR DESCRIPTION
**Pull request by pbalcer**
This fixes the implementation of UR_PROGRAM_INFO_DEVICES and UR_PROGRAM_INFO_NUM_DEVICES when using multi-devices programs.

### CODE REVIEW

**Comment by pbalcer** [comment]
https://github.com/intel/llvm/pull/12813

File: source/adapters/level_zero/program.cpp [hunk 1]
```diff
@@ -601,12 +601,16 @@ UR_APIEXPORT ur_result_t UR_APICALL urProgramGetInfo(
...
+  case UR_PROGRAM_INFO_NUM_DEVICES: {
+    return ReturnValue(static_cast<uint32_t>(Program->ZeModuleMap.size()));
+  }
+  case UR_PROGRAM_INFO_DEVICES: {
+    std::vector<ze_device_handle_t> devices;
+    for (const auto &entry : Program->ZeModuleMap) {
+      devices.push_back(entry.first);
+    }
+    return ReturnValue(devices.data(), devices.size());
+  }
```
Comment (omarahmed1111) [inline source/adapters/level_zero/program.cpp:613]: I think that should report the devices inside the context. From what I could find `ZeModuleMap` will only be accumulated when either `urProgramBuildExp` is called or `urProgramLinkExp` and we could call this query before calling these entries (I might be missing something here).

Comment (pbalcer) [reply source/adapters/level_zero/program.cpp]: Yea, this was wrong :D My fault for rushing.
From what I can see, all paths to build a program go through the exp functions except for urProgramCreateWithBinary. In which case `ZeModuleMap` is going to be empty.

Thanks.

**Review by nrspruit — APPROVED** [review]

**Pull request closed without merging**